### DISCUSSION TITLE
Getting Laravel-Sluggable to work with Spatie Laravel Tags

### DISCUSSION THREAD

**Discussion by Rocketpilot** [spatie/laravel-sluggable#282] [Q&A]
Apologies for reposting this - I originally had it in Laravel Tags' discussion area but that appears to be a ghost town right now.

I've been having trouble getting routing via slug working with Laravel Tags. I haven't created a custom Tags model to add a Sluggable trait, which would probably work, but the Tags docs don't give a complete example and I'm not an advanced enough developer to fill in the missing pieces. I have a feeling that shouldn't be required as they're both packages by the same developer.

I see there's a config setting for choosing which slugger to use, but again the documentation stops before it spells out how to do it.

**Comment by ismaildasci** [comment]
Hey! 👋

It sounds like you want to use the **Laravel-Sluggable** package with **Spatie Laravel Tags** to generate slugs for tags, but you're running into some issues due to limited documentation. Let’s go through a straightforward way to make them work together without needing to create a fully custom `Tags` model.

### Step-by-Step Guide

1. **Extend the Spatie Tag Model**:
   Instead of creating a brand new `Tags` model, you can extend Spatie’s `Tag` model to add the `Sluggable` trait.

   First, create a new model in your app’s models directory:
   ```php
   // app/Models/CustomTag.php
   namespace App\Models;

   use Illuminate\Database\Eloquent\Model;
   use Spatie\Tags\Tag;
   use Cviebrock\EloquentSluggable\Sluggable;

   class CustomTag extends Tag
   {
       use Sluggable;

       /**
        * Return the sluggable configuration array for this model.
        *
        * @return array
        */
       public function sluggable(): array
       {
           return [
               'slug' => [
                   'source' => 'name',  // The field you want to generate the slug from
               ]
           ];
       }
   }
   ```

   Here, we’re using the `name` attribute to generate the slug, which is typical for tag slugs.

2. **Update the Spatie Tags Configuration**:
   In your `config/tags.php` file, tell Spatie Tags to use your new `CustomTag` model instead of the default `Tag` model. 

   ```php
   // config/tags.php
   'tag_model' => App\Models\CustomTag::class,
   ```

   This will ensure that whenever Spatie creates or interacts with tags, it uses your extended model with the `Sluggable` trait.

3. **Add Slug Routing in Your Controller or Routes**:
   Now, you can use slugs in your routes. For example, you might set up a route to display posts by tag:

   ```php
   // routes/web.php
   use App\Models\CustomTag;

   Route::get('/tags/{slug}', function ($slug) {
       $tag = CustomTag::where('slug', $slug)->firstOrFail();
       $posts = $tag->posts; // assuming a relationship between tags and posts

       return view('tagged-posts', compact('tag', 'posts'));
   });
   ```

4. **Migrate Existing Tags** (If Needed):
   If you already have existing tags without slugs, you might need to generate slugs for them. You can do this with a simple artisan command or a script:

   ```php
   // In a command or a Tinker session
   App\Models\CustomTag::all()->each(function ($tag) {
       $tag->slug = null;  // Reset the slug so it regenerates
       $tag->save();
   });
   ```

5. **Test It**:
   Now, whenever you create a tag, the `slug` field should automatically populate based on the `name`, thanks to the `Sluggable` trait. You can access tags by their slug in the URL and link directly to them.

### Summary

1. Extend `Spatie\Tags\Tag` with a new `CustomTag` model and add `Sluggable`.
2. Update `config/tags.php` to use `CustomTag`.
3. Add routing to use slugs in URLs.
4. Regenerate slugs for existing tags if needed.

This approach avoids modifying the original Spatie package and leverages Laravel-Sluggable to handle slugs automatically. Let me know if this helps or if you have any further questions! 😊
